File: backend/schemas/vaep.py
"""
Pydantic schemas for VAEP and PlayeRank endpoints.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# ── VAEP Schemas ─────────────────────────────────────────────────────

class VAEPPipelineRequest(BaseModel):
    competition_id: Optional[int] = None
    season: str = Field(default="current", max_length=20)


class VAEPRating(BaseModel):
    player_name: str
    team: Optional[str] = None
    league: Optional[str] = None
    position: Optional[str] = None
    minutes_played: int = 0
    total_vaep: float = 0.0
    vaep_per90: float = 0.0
    offensive_vaep: float = 0.0
    defensive_vaep: float = 0.0
    actions_count: int = 0
    season: Optional[str] = None


class VAEPPipelineResponse(BaseModel):
    season: str
    method: str
    total_players: int
    total_actions: int
    total_games: int
    top_players: List[VAEPRating]


class VAEPRatingsResponse(BaseModel):
    total: int
    season: Optional[str] = None
    ratings: List[VAEPRating]


class VAEPActionDetail(BaseModel):
    action_type: str
    vaep_value: float
    offensive_value: float = 0.0
    defensive_value: float = 0.0
    x_start: Optional[float] = None
    y_start: Optional[float] = None
    x_end: Optional[float] = None
    y_end: Optional[float] = None
    minute: Optional[int] = None
    second: Optional[int] = None


class VAEPPlayerDetail(BaseModel):
    player_name: str
    team: Optional[str] = None
    league: Optional[str] = None
    position: Optional[str] = None
    minutes_played: int = 0
    total_vaep: float = 0.0
    vaep_per90: float = 0.0
    offensive_vaep: float = 0.0
    defensive_vaep: float = 0.0
    actions_count: int = 0
    top_actions: List[VAEPActionDetail] = []
    season: Optional[str] = None


class VAEPComparisonResponse(BaseModel):
    players: List[VAEPRating]


# ── PlayeRank Schemas ────────────────────────────────────────────────

class PlayeRankScore(BaseModel):
    player_name: str
    team: Optional[str] = None
    league: Optional[str] = None
    position: Optional[str] = None
    role_cluster: str = "unknown"
    composite_score: float = 0.0
    dimensions: Dict[str, float] = {}
    percentile_in_cluster: float = 0.0
    cluster_size: int = 0
    season: Optional[str] = None


class PlayeRankRankingsResponse(BaseModel):
    total: int
    role_cluster: Optional[str] = None
    season: Optional[str] = None
    rankings: List[PlayeRankScore]

File: frontend/src/types/vaep.ts
export interface VAEPRating {
  player_name: string;
  team: string | null;
  league: string | null;
  position: string | null;
  minutes_played: number;
  total_vaep: number;
  vaep_per90: number;
  offensive_vaep: number;
  defensive_vaep: number;
  actions_count: number;
  season: string | null;
}

export interface VAEPPipelineResponse {
  season: string;
  method: string;
  total_players: number;
  total_actions: number;
  total_games: number;
  top_players: VAEPRating[];
}

export interface VAEPRatingsResponse {
  total: number;
  season: string | null;
  ratings: VAEPRating[];
}

export interface VAEPActionDetail {
  action_type: string;
  vaep_value: number;
  offensive_value: number;
  defensive_value: number;
  x_start: number | null;
  y_start: number | null;
  x_end: number | null;
  y_end: number | null;
  minute: number | null;
  second: number | null;
}

export interface VAEPPlayerDetail extends VAEPRating {
  top_actions: VAEPActionDetail[];
}

export interface VAEPComparisonResponse {
  players: VAEPRating[];
}

export interface PlayeRankScore {
  player_name: string;
  team: string | null;
  league: string | null;
  position: string | null;
  role_cluster: string;
  composite_score: number;
  dimensions: Record<string, number>;
  percentile_in_cluster: number;
  cluster_size: number;
  season: string | null;
}

export interface PlayeRankRankingsResponse {
  total: number;
  role_cluster: string | null;
  season: string | null;
  rankings: PlayeRankScore[];
}
